src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdateHarvestedDatasetCommand.java

```diff
@@ -0,0 +1,201 @@
+
+    public UpdateHarvestedDatasetCommand(Dataset theDataset, DatasetVersion newHarvestedVersion, DataverseRequest aRequest) {
+        super(aRequest, theDataset);
+        this.newHarvestedVersion = newHarvestedVersion;
+    }
+
+    public boolean isValidateLenient() {
+        return validateLenient;
+    }
+
+    @Override
+    public Dataset execute(CommandContext ctxt) throws CommandException {
+        
+        Dataset existingDataset = getDataset();
+
+        if (existingDataset == null
+                || existingDataset.getId() == null
+                || !existingDataset.isHarvested()
+                || existingDataset.getVersions().size() != 1) {
+            throw new IllegalCommandException("The command can only be called on an existing harvested dataset with only 1 version", this);
+        }
+        DatasetVersion existingVersion = existingDataset.getVersions().get(0);
+
+        if (newHarvestedVersion == null || newHarvestedVersion.getId() != null) {
+            throw new IllegalCommandException("The command can only be called with a newly-harvested, not yet saved DatasetVersion supplied", this);
+        }
+        
+        newHarvestedVersion.setCreateTime(getTimestamp());
+        newHarvestedVersion.setLastUpdateTime(getTimestamp());
+
+        
+        Map<String, Integer> existingFilesIndex = new HashMap<>();
+
+        /* 
+        Create a map of the files that are currently part of this existing
+        harvested dataset. We assume that a harvested file can be uniquely 
+        defined by its storageidentifier. Which, in the case of a datafile
+        harvested from another Dataverse should be its data access api url.
+        */
+        for (int i = 0; i < existingDataset.getFiles().size(); i++) {
+            String storageIdentifier = existingDataset.getFiles().get(i).getStorageIdentifier();
+            if (storageIdentifier != null) {
+                existingFilesIndex.put(storageIdentifier, i);
+            }
+        }
+        
+        /*
+        Go through the files in the newly-harvested version and check if any of 
+        them are (potentially new/updated) versions of files that we already
+        have, harvested previously from the same archive location. 
+        */
+        for (FileMetadata newFileMetadata : newHarvestedVersion.getFileMetadatas()) {
+            // is it safe to assume that each new FileMetadata will be 
+            // pointing to a non-null DataFile here?
+            String location = newFileMetadata.getDataFile().getStorageIdentifier();
+            if (location != null && existingFilesIndex.containsKey(location)) {
+                newFileMetadata.getDataFile().setFileMetadatas(new ArrayList<>());
+
+                int fileIndex = existingFilesIndex.get(location);
+                
+                // Make sure to update the existing DataFiles that we are going 
+                // to keep in case their newly-harvested versions have different 
```

Comment:
Not sure I'm following the overall goal - If these are different are you assuring their reindexed and that they aren't otherwise?

Reply:
The goal is to end up with a harvested dataset/datafiles in the database that match the original objects on the source side as closely as possible, while preserving the database ids of the previously harvested objects as much as possible.

Reply:
OK - I guess my underlying concern is to make sure we don't over-optimize in indexing and assume that if you have the same fileMetadata and datafile ids that we don't have to check for changes. Not sure we do that now, but not changing existing published ones here would avoid the problem as well. (I don't think Curate reuses ids for fileMetdatas - perhaps it could if they haven't changed at all to do the same optimization (although I understand the issue for Harvest is doing many datasets all at once.)

Reply:
May be obvious by now, but my line of thought was, since our solr documents are based on db ids, if I preserve whatever db ids I can, we can, at least in theory, save on deleting the old solr docs and re-creating them from scratch; and instead modify those existing docs. I am actually not 100% sure that that's what our indexing code currently does when reindexing existing datasets. (?) But at least it'll be possible to achieve this, if the solr doc names stay the same for datasets and datafiles that have already been harvested before. 

I was not yet thinking of/planning to able to skip the indexing of the datasets and datafiles that haven't changed since the last harvest. (I don't think we can, in any economical way - ?). But I am hoping that not having to recreate solr docs unnecessarily should be a benefit. 

Case in point, we have 20K datasets harvested from Borealis. With files it's 300K dvobjects, plus the same number of perm. docs - on the order of magnitude of 1M. They have marked most of the OAI records as having been updated since the last time we harvested. So then we are talking of deleting and rebuilding from scratch that many docs on an index that has close to 10M docs total. Being able to reuse most of them gotta be a serious saving - ?